Output Unicode Hangul syllables directly, to eliminate need for Korean keyboard layout

gfixler opened the pull request:
안녕하세요, Peter!

Sorry for this pull request out of nowhere, but I saw in the "How you can contribute" section that you were looking to bypass the "system keyboard layout" work around, and it seemed like fun. I updated it to use some arithmetic on the initial, medial, and final bits, to find, and directly output the Unicode Hangul syllable. Seems to work; I wrote the greeting above using it. I explained everything in the few commit messages. I don't know Korean, so I did a lot of reading last night, and reverse-engineering of the code and system. The two remaining things I don't get/have working yet are "tense" final consonants (can they even occur in final position?), and compound final consonants. Do you have any more info about those?

It was really fun to look up Korean words last night, and start to be able to write some of them out, first try. Hangul feels a lot like steno. Totally fine if you're not interested in this pull request!

Commit:
Replace Dubeolsik system with Unicode translation

The previous system output English characters, which would then be
converted to Korean letters and syllables through the computer's Korean
keyboard layout. This required the user to switch over to the Korean
layout first, before using this plugin.

The new system uses some arithmetic, and numbers assigned to all the
letters, to determine which of the ~11,000 Hangul syllable block code
points to return. See the full table of Hangul syllables here:

https://en.wikipedia.org/wiki/Hangul_Syllables

Here's how the algorithm works...

19 initial jamo:

0. ᄀ 4. ㄸ  8. ㅃ 12. ㅈ 16. ㅌ
1. ㄲ 5. ㄹ  9. ㅅ 13. ㅉ 17. ㅍ
2. ㄴ 6. ㅁ 10. ㅆ 14. ㅊ 18. ㅎ
3. ㄷ 7. ㅂ 11. ㅇ 15. ㅋ

21 medial jamo:

0. ㅏ  7. ㅖ 14. ㅝ
1. ㅐ  8. ㅗ 15. ㅞ
2. ㅑ  9. ㅘ 16. ㅟ
3. ㅒ 10. ㅙ 17. ㅠ
4. ㅓ 11. ㅚ 18. ㅡ
5. ㅔ 12. ㅛ 19. ㅢ
6. ㅕ 13. ㅜ 20. ㅣ

27 final jamo (28 including "None"):

0.     7. ㄷ 14. ㄿ 21. ㅇ
1. ㄱ  8. ㄹ 15. ㅀ 22. ㅈ
2. ㄲ  9. ㄺ 16. ㅁ 23. ㅊ
3. ㄳ 10. ㄻ 17. ㅂ 24. ㅋ
4. ㄴ 11. ㄼ 18. ㅄ 25. ㅌ
5. ㄵ 12. ㄽ 19. ㅅ 26. ㅍ
6. ㄶ 13. ㄾ 20. ㅆ 27. ㅎ

We plug those numbers into this formula:

44032 + initial * 588 + medial * 28 + final

The first number gets us to the start of the Hangul set of characters in
the Unicode system, and the rest offset us from there to the syllable
that matches the parts. The 28 final characters repeat over and over,
for each medial, and the medials repeat every 21 medials * 28 finals, or
every 588 characters, and each of those repeats starts with a different
initial character.

Currently, I don't now how to slot in the "tense" final consonants—I
just don't understand how those work in Hangul, so their contributions
are currently set to 0—and I can't see how the compound final consonants
are meant to be created. For the latter, I watched the demo video,
explored the code, and did some looking around online, and never really
saw them in use anywhere. When I understand these things, I can update
the code.

Commit:
Remove comment ability Korean keyboard layout need

We no longer need to switch the computer keyboard to the Korean layout.

Commit:
Let Plover handle the star key

In experimenting with this system, if I misstroked, I'd be passed
through to other dictionaries below this one, outputting whatever
matched there, but then this dictionary would capture all "*" strokes,
and always want to move left, right, then backspace. This meant that
after a misstroke, I would be stuck backspacing character-by-character.
Passing star back out to Plover means star works more as expected, and
allows things like writing numbers among the Hangul, while still being
able to easily star back. This doesn't disrupt deletion of the Korean
syllables, unless the Dubeolsik keyboard system also allowed backspacing
through each individual jamo, which this doesn't.

petercpark commented:
This is very promising! I'm very impressed with your work. I tested it out just now and it works very well.
Yes we need compound final consonants for words like "a lot" (많이) or "chicken" (닭).

Apparently there are 11 total: ㄳ, ㄵ, ㄶ, ㄺ, ㄻ, ㄼ, ㄽ, ㄾ, ㄿ, ㅀ, and ㅄ
Here's how I would define them:
ㄳ, GS
ㄵ, PBG
ㄶ, FPB
ㄺ, RG and LG
ㄻ, RPL
ㄼ, RB
ㄽ, RS and LS
ㄾ, RT and LT
ㄿ, RP
ㅀ, FR and FL
ㅄ, BS and BZ

double final consonants are also necessary. For example "there is" (있다) and "to saute" (볶다).
I think there is only double SS and double GG. Other doubles are never used for ending consonants.
ㅆ, SZ and ST
ㄲ, GZ and GT

Also, it would be good to include single letter outputs to enable things like ㅋㅋㅋ (laughing, in context of text messages). Single characters would include all letters, tense letters, and vowels. I think originally any character on the left hand or vowels would output as a single character if not paired with anything else.

Commit:
Reorganize consonants and vowels dictionaries

Now that the jamos are numbered, to work with the arithmetic of how
they're laid out in Unicode, I think it makes sense to order them that
way in the lists, so it's easier to see what exists, or is missing.

Commit:
Fix ending tense consonants

It turns out there are only two that get doubled at the end of
syllables. These don't require the earlier addition of an R. Thanks to
project creator, Peter, for the chord choices.

Commit:
Add all ending compound consonants

These were missing in this plugin, but they all work with the formula in
which Unicode has all syllables laid out, so they were easy to add.
Thanks to project creator, Peter, for the chords used by these.

Commit:
Pair each jamo offset with its jamo

This moves the jamo comments after each dictionary entry into a pair
with their offset integers, reifying them as data we can use. This is in
preparation for being able to output each on its own.

Commit:
Justify dictionary entries

This is totally unnecessary, but very orderly.

Commit:
Break up the starting consonants dictionary

This simplifies things a bit, and lines things up, too, for no very
important reason. Also added some blank lines to make it easier to
navigate sections (I'm a Vim guy).

Commit:
Fix some errors in vowel entries

A TH variant chord for ㄸ was accidentally in the regular consonants
dictionary. The O chord for ㅗ was in the regular and y-vowel
dictionaries, but the OE variant was only in the regular, where it
denotes the ㅛ vowel. There was an incorrect AOU entry for ㅠ, which
should only work for ㅡ. Finally, the entry for "EU"=ㅣwas missing.

petercpark commented:
Okay I'm just reading through your code, and I think I found out a way to implement what we're talking about. It requires using Jamo Unicode instead of entire syllable. I haven't written the code yet, I'm going to test to see if it works.

Commit:
Add the two missing vowels, as comments

There aren't enough vowel chords for all of the vowels in Korean, even
moving y-vowels to the left-hand side consonants. The two commented-out
are the least, and third-least frequently used vowels, according to:
https://www.kref.altervista.org/kreffrequency.html

gfixler commented:
Sorry, I only just saw your comment. It's been a busy few days. I made several more commits. I got the things we discussed above working. They were just missing entries in the tables. I also moved the Korean letters from comments after each letter into tuples with them, so they could pull double duty as showing what letter they were in the code, and being able to write each letter through the system. I've almost got that working locally. Many do work, but I need to think through some edge cases. I put some explanatory stuff in each commit message.

Commit:
Begin allowing single letter output

This still needs work, going through the logic. The initial consonants
work on their own, as do [most of?] the vowels, but the final consonants
output entire syllables, combined with 아.

gfixler commented:
Added latest, which allows outputting initial consonants or vowels on their own, even the star/y-vowels, but not yet final consonants, and I think there may still be an issue with some of the vowels.

Commit:
Remove duplicate vowel

AOU was being used for ㅝ and ㅡ, and since ㅡ already has a chord, AU,
I removed the AOU → ㅡ mapping. In the process, I also realized that
we're not currently using AOE for anything. We're using 14 of the 15
vowel chords (not including the 16th chord, which is no keys pressed).
This means we could give that chord to one of the two remaining vowels
that doesn't have a chord. We could also give the lack of any chord keys
to the final chord, and then have them all. We wouldn't be able to write
that chord on its own, but that could be filled in with a custom
sequence, or we could just give it to the least common one.

I also removed the empty case for starting vowels, which I think is
right, but will need some more testing to confirm.

gfixler commented:
Okay a few discoveries...

1. We had a dupe in there. AOU was mapped to both ㅝ and ㅡ. As the latter had another chord in there, AU, I removed the AOU → ㅡ mapping. You'll have to let me know if that's correct.
2. We have an extra chord to play with. We're not yet using AOE. If it maps well to ㅙ or ㅞ, then we can check one of the two still-unmapped vowels off the list.
3. I also realized we could use the skeleton chords, i.e. no vowel keys, for the final one. It'll have to be up to you how the steno keys map, but if you can look back over the vowels, and figure out if AOE, and the no-vowels chord, work with the rest, maybe even moving others around for a best fit, then we'd have everything.

Well... We'd have everything, except for being able to write the skeleton chord vowel on its own, but if it were, say, ㅒor ㅞ, which, according to [this page](https://www.kref.altervista.org/kreffrequency.html), are the least used, it may not matter too much. It could always use a hardcoded outline.

gfixler commented:
Found another duplicate. In the ending consonants, -PBG is mapped to both ㅇ and ㄵ. How to fix?

Commit:
Fix lone ending consonant usage

Whoops, typo. Now we have ending consonants that can be used on their
own. Stroke the starting TKPW chord, and it outputs ㄱ. Stroke the AOU
chord, and you get ㅝ. Stroke the ending -FR chord, you get ㅀ.

If you want a vowel that can have Y added, like ㅏ → ㅑ, just add the
star key, so A is ㅏ, and A* is ㅑ. This is requirement is due to the
fact that adding the Y sound is normally done as part of the initial
consonant, as in KA → 가, vs KWA → 갸, so we can't write those vowels on
their own otherwise.

gfixler commented:
Okay, I knew not being able to write the final consonants would be something dumb. I had the wrong variable name in there 🙄. Fixed! Now we can write...

initial consonants:
`TKPW` → ㄱ
`SKWR` → ㅈ
`KP` → ㅋ

tense initial consonants:
`S*` → ㅆ
`K*` → ㄲ
`PW*` → ㅃ

vowels:
`A` → ㅏ
`AU` → ㅡ
`OEU` → ㅚ

y-vowels:
`A*` → ㅑ
`O*E` → ㅛ
`A*EU` →ㅒ

final consonants:
`-PL` → ㅁ
`-FR` → ㅀ
`-GT` → ㄲ

initial consonant + vowel pairs:
`KHA` → 차
`PWOE` → 보
`TPHAOU` → 눠

tense initial consonant + vowel pairs:
`S*U ` → 써
`PWO*EU` → 뾔
`SKWRA*U` → 쯔

initial consonant + y-vowel pairs:
`KWA` → 갸
`SHOE` → 쇼
`KWRO` → 요

tense initial consonant + y-vowel pairs:
- this pairing currently fails
`KWA*` should be ㄲ +ㅑ, but yields 갸 instead.

full syllables:
`KAT` → 같
`TKOG` → 독
`PWAEUG` → 백

Commit:
Add TH variant of ㄷ(TK) to regular, special dicts

This is a variant in the tense dictionary, but wasn't in the regular and
special versions of the starting consonants.

Commit:
Add combined version of special, tense systems

The way things are currently laid out, the easiest way to get this last
piece of the puzzle in place was to create a tense+special dictionary,
which combines the chords of the special dict with the doubled letters
of the tense dict. This seems to work now, so using a special chord, say
SH instead of S—to get the version of an initial consonant with a
y-vowel after it—along with the star—to get the tense (doubled) version
of that consonant—will finally output both things together in the
syllable. That means all the things work. There's just a little bit of
cleanup and vowel decisions remaining.

gfixler commented:
Okay, got it all working (I think). The "tense initial consonant + y-vowel pairs" mentioned in the previous comment, which wasn't working, now works. It adds another dictionary, and another branch in the initial consonant `if` block, but it's fairly straightforward.

I also added the TH variant of the TD chord from the tense initial consonant dictionary to the regular and special initial consonant dictionaries, because it was missing from those, but need confirmation that this variant should also be in those.

Commit:
Switch from attach operators to glue operator

The attach operators ({^}example{^}) forced syllables to stick to
everything on both sides. This ({&example}) makes them stick only to
other glued things, like each other, so if—using KR-PB as my Korean
dictionary toggle—I stroke:

-T/WORD/KR-PB/TKPWA/SKWROBG/KR-PB/PHAOEPBS/TPAEPL

...with the attach operators I would get:

the word가좈means family

...but with the glue, I get:

the word 가좈 means family
